### ISSUE TITLE
Provide asynchronous/preloading navigation API

### ISSUE THREAD

**Issue by tbuschto** [eclipsesource/tabris-js#797]
Maybe this is a topic for 2.0 since navigation is a very central concept, but let me describe the problem anyway:

It seems to be a recurring issue that when a page is created and opened, there are still some tasks to perform before the UI is in its final state. That includes loading images, webviews, or displayable data via asynchroneous http or file APIs. This results in the ugly scenario the that the new UI changes during or shortly after the page transition. The same issue exists - to a smaller degree - for the initial UI not part of a page (navigation and drawer elements).

A limited workaround would be to create the page, wait for a synthetic load event, then open it. However, this doesn't work for image pre-loading (we would need load events on all widgets that can display images, and at least in windows an image doesn't start loading until the widget is connected to the UI tree), and it delays the page transition, making the ui less responsive. It's also a lot of work.

I can only speak from the perspective of a windows developer, but there is a solution that can automtically wait (with a given timeout) for all images to load - which would already be a huge win - before starting the intro animation for the new page. It also has an optional hook to indicate to the navigation logic when the page is ready to be displayed. If we provided such features as API it could fix all the issues described above.

**Comment by mpost** [comment]
What kind of images are you referring to? Bundled images or images loaded remotely?

**Comment by tbuschto** [comment]
Both actually, since they are both loaded async (at least on windows, but remote images are loaded async on all platforms I would assume).

**Comment by mpost** [comment]
Yeah images are always loaded async everywhere. I would recon that local loading is sufficiently fast though.

I would hesitate to include any kind of artificial delay. The ui should make appropriate transitions as soon as possible in order to give the feeling of a snappy ui.

**Comment by mpost** [comment]
Adding to the above comment what could be interesting is to load images independently from the ImageView. Eg. preloading when you know you will need it next. Like heating up the cache.

In some future where we would be able to properly fetch images and store them on disc that could also be implemented in user code.

**Comment by ralfstx** [comment]
I think displaying contents as they become available isn't necessarily an "ugly" effect. In most cases I suppose it will be preferable to delaying the initial UI.

If an application wants to wait for content being loaded, shouldn't that be possible using the `load` events of `WebView` and `ImageView`?

**Comment by tbuschto** [comment]
It's pretty ugly when the image changes the layout (which will happen whenever you rely on its intrinsic size) or covers a bigger area of the new page. I also think the "thumbs-up" appearing after a delay in the tabris-con app is pretty ugly.

Just to explain how it works in windows: There is a page-outro and a page-intro animation which play in succession. You can make it so that the page outro is played immediately, but the page intro is played only once the new page is ready. In-between that the area is blank (white or black, with a spinner if you like). That's what I would like to offer as an option, and you can't do that using only load events or pre-caching, especially if the load events rely on the widget being in the widget-tree.

A similar issue exists with the splash screen in windows (it may go away to soon), but that is much harder to fix.

**Comment by mpost** [comment]
@tbuschto Is the problem that the image loading takes place to late? Eg. only when the page is fully visible (after the intro animation).

**Comment by tbuschto** [comment]
After or during the animation, that's the scenario. I have to admit though that this is not as often the case at it first seemed since we often use the remote script loading feature. When packaged it's mostly only a problem when loading resources from elsewhere.

Ideas:

Using magical blocking API:
`page.preload().open();` // could probably not handle XHR though....

Using promise:
`page.preload().then(=> page.open());`

Less magical using promise and two-step opening:
`page.attach().find("Image").when("load").then(=> page.open());`

Implicit two-step opening by giving a Promise (seems the best idea to me):
`page.open(page.find("Image").when("load"));`
